solr/core/src/java/org/apache/solr/handler/component/HttpShardHandler.java

```diff
@@ -232,13 +237,16 @@ public ShardResponse takeCompletedIncludingErrors() {
    */
   @Override
   public ShardResponse takeCompletedOrError() {
-    return take(true);
+    return take(true, Long.MAX_VALUE);
   }
 
-  private ShardResponse take(boolean bailOnError) {
+  private ShardResponse take(boolean bailOnError, long maxAllowedTime) {
     try {
       while (pending.get() > 0) {
-        ShardResponse rsp = responses.take();
+        long waitTime = maxAllowedTime - System.nanoTime();
+        ShardResponse rsp = responses.poll(waitTime, TimeUnit.NANOSECONDS);
```

Comment:
This is not quite doing what we want here.

`maxAllowedTime` is supplied in millis, so any computations need to make sure the values are compatible. At the head of this method we want e.g. to convert
```java
long maxAllowedNanos = TimeUnit.MILLISECONDS.toNanos(maxAllowedTime);
long startTimeNanos = System.nanoTime();
```
then compute wait time for each poll:
```java
long waitTimeNanos = maxAllowedNanos - (System.nanoTime() - startTimeNanos);
```